based on https://chromium.googlesource.com/chromiumos/third_party/kernel/+/365a4da19d1bfa95a5f213620cdc70dec920ef8f%5E%21/#F2

there are more chromeos arm errata fixes, which seem not to be in mainline yet:
773769 - https://chromium.googlesource.com/chromiumos/third_party/kernel/+/cb3f3fcb2becf8c58588047c271e5e2f8aa9896e%5E!/#F2.patch
774769 - https://chromium.googlesource.com/chromiumos/third_party/kernel/+/72128039f77489b04c0eb241864cea87d9e3ed9e%5E%21/#F2

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 002e0cf025f5..a91389ec36f4 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -1120,6 +1120,18 @@ config ARM_ERRATA_857272
 	  config option from the A12 erratum due to the way errata are checked
 	  for and handled.
 
+config ARM_ERRATA_766421
+       bool "ARM errata: Strongly-Ordered/Device load or NC LDREX could return incorrect data"
+       depends on CPU_V7
+       help
+         This option enables the workaround for the 766421 Cortex-A15 erratum.
+         In certain situations, a strongly ordered or device load instruction,
+         or a non-cacheable normal memory load-exclusive instruction could
+         match multiple fill buffers and return incorrect data.
+         This workaround is add DMB instruction when making any change to the
+         translation regime and before doing any new loads/stores/preloads
+         in the new translation regime.
+
 endmenu
 
 source "arch/arm/common/Kconfig"
diff --git a/arch/arm/mach-exynos/Kconfig b/arch/arm/mach-exynos/Kconfig
index d2d249706ebb..8c61dce83cbc 100644
--- a/arch/arm/mach-exynos/Kconfig
+++ b/arch/arm/mach-exynos/Kconfig
@@ -70,6 +70,7 @@ config ARCH_EXYNOS4
 config ARCH_EXYNOS5
 	bool "Samsung Exynos5"
 	default y
+	select ARM_ERRATA_766421
 	help
 	  Samsung Exynos5 (Cortex-A15/A7) SoC based systems
 
diff --git a/arch/arm/mm/proc-v7-2level.S b/arch/arm/mm/proc-v7-2level.S
index 5db029c8f987..47a4e6bd8988 100644
--- a/arch/arm/mm/proc-v7-2level.S
+++ b/arch/arm/mm/proc-v7-2level.S
@@ -50,6 +50,8 @@ ENTRY(cpu_v7_switch_mm)
 #endif
 #ifdef CONFIG_ARM_ERRATA_754322
 	dsb
+#elif CONFIG_ARM_ERRATA_766421
+	dmb
 #endif
 	mcr	p15, 0, r1, c13, c0, 1		@ set context ID
 	isb
